File: MoonEngine-Root/Mesh/src/main/java/com/greentree/engine/moon/mesh/compoent/NormalStaticMeshVertexComponent.java
package com.greentree.engine.moon.mesh.compoent;

import com.greentree.commons.math.vector.Vector3f;
import com.greentree.engine.moon.mesh.StaticMesh.MeshFace;
import com.greentree.engine.moon.mesh.StaticMesh.VertexIndex;


public final class NormalStaticMeshVertexComponent implements StaticMeshVertexComponent {
	
	public static final NormalStaticMeshVertexComponent INSTANCE = new NormalStaticMeshVertexComponent();
	
	private NormalStaticMeshVertexComponent() {
	}
	
	@Override
	public void get(MeshFace face, int strip, float[] a, float[] b, float[] c) {
		if(face.hasNormals())
			StaticMeshVertexComponent.super.get(face, strip, a, b, c);
		else {
			var va = face.a().vertex();
			var vb = face.b().vertex();
			var vc = face.c().vertex();
			
			final var normal = va.sub(vb, new Vector3f()).cross(vc.sub(vb, new Vector3f()))
					.normalize(-1);
			
			StaticMeshFaceComponent.setVector(a, strip, normal);
			StaticMeshFaceComponent.setVector(b, strip, normal);
			StaticMeshFaceComponent.setVector(c, strip, normal);
		}
	}
	
	@Override
	public void get(VertexIndex index, int strip, float[] dest) {
		if(!index.hasNormal())
			throw new UnsupportedOperationException();
		StaticMeshFaceComponent.setVector(dest, strip, index.normal());
	}
	
	@Override
	public int size() {
		return 3;
	}
	
}

File: MoonEngine/Mesh/src/main/java/com/greentree/engine/moon/mesh/compoent/NormalStaticMeshVertexComponent.kt
package com.greentree.engine.moon.mesh.compoent

import com.greentree.engine.moon.mesh.StaticMesh
import com.greentree.engine.moon.mesh.StaticMesh.*

data object NormalStaticMeshVertexComponent : StaticMeshVertexComponent {

	override fun get(face: MeshFace, strip: Int, a: FloatArray, b: FloatArray, c: FloatArray) {
		if(face.hasNormals()) super.get(face, strip, a, b, c) else {
			val va = face.a.vertex()
			val vb = face.b.vertex()
			val vc = face.c.vertex()
			val normal = (va - vb).cross(vc - vb)
				.normalize(-1)
			StaticMeshFaceComponent.setVector(a, strip, normal)
			StaticMeshFaceComponent.setVector(b, strip, normal)
			StaticMeshFaceComponent.setVector(c, strip, normal)
		}
	}

	override fun get(index: VertexIndex, strip: Int, dest: FloatArray) {
		if(!index.hasNormal()) throw UnsupportedOperationException()
		StaticMeshFaceComponent.setVector(dest, strip, index.normal())
	}

	override fun size(): Int {
		return 3
	}
}